browser/net/brave_oauth_protocol_handler.cc

```diff
@@ -0,0 +1,232 @@
+  }
+
+  const std::string scheme = url.scheme();
+
+  if (!preserve_path) {
+    return GURL(redirects.at(scheme));
+  }
+
+  std::string query;
+  std::string path = url.path();
+
+  if (url.has_query()) {
+    query = base::StrCat({
+      "?",
+      net::EscapeExternalHandlerValue(url.query())
+    });
+  }
+
+  base::ReplaceFirstSubstringAfterOffset(&path, 0, "/", "");
+
+  return GURL(
+      base::StrCat({
+        redirects.at(scheme),
+        path,
+        query
+      }));
+}
+
+void SetAuthToken(const std::string& auth_token,
+                  const std::string& oauth_scheme,
+                  content::WebContents* web_contents) {
+  if (auth_token.empty()) {
+    return;
+  }
+
+  if (oauth_scheme == kBinanceScheme) {
+#if BUILDFLAG(BINANCE_ENABLED)
+    Profile* profile =
+        Profile::FromBrowserContext(
+          web_contents->GetBrowserContext());
+    BinanceServiceFactory::GetInstance()
+      ->GetForProfile(profile)
+      ->SetAuthToken(auth_token);
+#endif
+    } else if (oauth_scheme == kGeminiScheme) {
+#if BUILDFLAG(GEMINI_ENABLED)
+    Profile* profile =
+        Profile::FromBrowserContext(
+          web_contents->GetBrowserContext());
+    GeminiServiceFactory::GetInstance()
+      ->GetForProfile(profile)
+      ->SetAuthToken(auth_token);
+#endif
+  }
+}
+
+bool IsAllowed(content::WebContents* web_contents,
+               const base::Optional<url::Origin>& initiating_origin,
+               const std::string& origin) {
+  GURL allowed_origin(origin);
+  bool unallowed_origin =
+    !initiating_origin.has_value() ||
+    initiating_origin.value().GetURL() != allowed_origin ||
+    web_contents->GetLastCommittedURL().GetOrigin() != allowed_origin;
+  return !unallowed_origin;
+}
+
+bool ShouldRedirect(content::WebContents* web_contents,
+                    const base::Optional<url::Origin>& initiating_origin,
+                    const std::string& scheme) {
+  if (scheme == kRewardsScheme) {
+    for (const auto& origin : rewards_origins) {
+      if (IsAllowed(web_contents, initiating_origin, origin)) {
+        return true;
+      }
+    }
+    return false;
+  }
+
+  return IsAllowed(web_contents, initiating_origin, origins.at(scheme));
+}
+
+void LoadNewTabURL(
+    const GURL& url,
+    content::WebContents::OnceGetter web_contents_getter,
+    ui::PageTransition page_transition,
+    bool has_user_gesture,
+    const base::Optional<url::Origin>& initiating_origin) {
+  content::WebContents* web_contents = std::move(web_contents_getter).Run();
+  if (!web_contents) {
+    return;
+  }
+
+  const auto ref_url = web_contents->GetURL();
+  if (!ref_url.is_valid()) {
+    return;
+  }
+
+  const std::string oauth_scheme = url.scheme();
+  if (!ShouldRedirect(web_contents,
+         initiating_origin, oauth_scheme)) {
+    return;
+  }
+
+  SetAuthToken(
+    GetAuthToken(url), oauth_scheme, web_contents);
+
+  const auto redirect_url = GetRedirectURL(url,
+    (oauth_scheme == kRewardsScheme));
```

Comment:
Honestly, I think we are adding more complexity than there was before. Previously, to add a new handler, one had to write a trivial self-contained function in their own module. Now, to add a handler I need to find this piece of code (unrelated to the specific module, hardcode 3 custom constants into it, patch `IsOauthProtocol`, patch `SetAuthToken` and probably add some custom logic (see the rewards case).

We also add very custom dependencies to `browser/net` which is not actually designed to be aware of super domain-specific components like binance or gemini.

So instead I propose:
1) Leave rewards alone, because this case is obviously doesn't fit here. We can simplify and shorten this handler a lot, for example by not putting all query params into the map while iterating over them.
2) Move all constants back to their modules
3) Add here a common function, that would be parameterized by redirect urls, acceptable initiators, etc and a callback for setting auth token
4) Call this function from brave content client with proper parameters obtained from proper modules